crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1665 @@
+                            let Ok(handle) = row_group_data_rx.recv().await else {
+                                is_finished.store(true, Ordering::Relaxed);
+                                break;
+                            };
+
+                            let row_group_data = handle.await?;
+
+                            let (df, morsel_seq) = row_group_decoder
+                                .row_group_data_to_df(row_group_data)
+                                .await?;
+
+                            // TODO: Split the df size if the row group is too large.
+
+                            let mut morsel = Morsel::new(df, morsel_seq, source_token.clone());
+                            morsel.set_consume_token(wait_group.token());
+
+                            if morsel_tx.send(morsel).await.is_err() {
+                                break;
+                            }
+
+                            wait_group.wait().await;
+
+                            if source_token.stop_requested() {
+                                break;
+                            }
+                        }
+
+                        Ok(row_group_data_rx)
+                    }));
+            }
+
+            assert!(self.row_group_data_receivers.is_empty());
+
+            join_handles.push(scope.spawn_task(TaskPriority::Low, async move {
+                for handle in self.spawned_task_handles.drain(..) {
+                    self.row_group_data_receivers.push(handle.await?);
+                }
+
+                if self.is_finished.load(Ordering::Relaxed) {
+                    self.shutdown().await?;
+                }
+
+                Ok(())
+            }));
+        }
+    }
+}
+
+impl ParquetSourceNode {
+    /// # Panics
+    /// Panics if called more than once.
+    async fn shutdown(&mut self) -> PolarsResult<()> {
+        if self.verbose {
+            eprintln!("[ParquetSource]: Shutting down");
+        }
+
+        self.row_group_data_receivers.clear();
+        // Join on the producer handle to catch errors/panics.
+        // Safety
+        // * We dropped the receivers on the line above
+        // * This function is only called once.
+        self.row_group_data_task_handle.take().unwrap().await
+    }
+
+    /// Constructs the task that fetches the bytes of row groups.
+    #[allow(clippy::type_complexity)]
+    fn init_row_group_data_fetcher(
+        &mut self,
+    ) -> (
+        Vec<
+            distributor_channel::Receiver<
+                async_executor::AbortOnDropHandle<PolarsResult<RowGroupData>>,
+            >,
+        >,
+        async_executor::AbortOnDropHandle<PolarsResult<()>>,
+    ) {
+        let verbose = self.verbose;
+        let io_runtime = polars_io::pl_async::get_runtime();
+
+        let use_statistics = self.options.use_statistics;
+
+        let (mut row_group_data_tx, row_group_data_receivers) = {
+            let (num_receivers, buffer_size_per_receiver) = if let DynByteSourceBuilder::Mmap =
```

Comment:
I reduced the number of receivers if the row group prefetch size to create backpressure - it would mean we wouldn't send morsels on some of the ports.

Would there be any issues if some of the ports don't receive any morsels? I'm currently changing the way backpressure is applied so I can make it 1:1 again if needed.

Reply:
> Would there be any issues if some of the ports don't receive any morsels? I'm currently changing the way backpressure is applied so I can make it 1:1 again if needed.

Yes, you'll lose parallelism capabilities in the entirety of the rest of the pipeline. Please make it 1:1.

Reply:
Updated to be 1:1